File: functions/segment_my.py
"""
Сегментация - поиск скачков скорости
"""
import numpy as np
from utils import setup_figure, finalize_figure
from config import config


def segment_my(Y, start_offset=1, window_shift=100, window_width=200,
               num_windows=200, fig=None):
    """
    Поиск скачков в данных.

    Returns
    -------
    J : np.ndarray
        Индексы скачков
    """
    Y = np.array(Y).flatten()
    s = config.plots.style
    threshold = config.constants.segment_threshold_mult

    Y1 = np.diff(Y)
    ss = np.std(np.abs(Y1))
    J = np.where(np.abs(Y1) > ss * threshold)[0]

    Y2 = np.zeros_like(Y1)
    Y2[J] = Y1[J]

    fig, [ax], standalone = setup_figure(fig, figsize=(12, 6))

    ax.plot(Y2, linewidth=s.thick_linewidth)
    ax.grid(True)
    ax.set_title('Скачки скорости',
                 fontsize=s.title_fontsize, fontweight=s.title_fontweight)
    ax.set_xlabel('Номер окна', fontsize=s.small_label_fontsize)

    finalize_figure(fig, standalone)
    return J

File: matlab/segment_my.m
function J=segment_my(Y,K,d,L,R);
n=length(Y);
Y1=diff(Y);                 %скорость
ss=std(abs(Y1));            %СКО 
J=find(abs(Y1)>ss*12);      JJ=size(J);   %номера скачков
Y2=zeros(size(Y1));
Y2(J)=Y1(J);

%Y3=[];
%for k=1:JJ;
%    j=J(k);
%    y=Y(j-L:j);
%    Y3=[Y3;y];
%end;

hF6=figure; 
set(hF6,'Units','Normalized','Pos',[0.2 0.175 0.78 0.55],...
    'Color',[0.8 0.8 0.8]);
plot(Y2,'LineWidth',4); grid;  %axis([0 n 0 1.2]);
title('Скачки скорости','FontSize',14,'FontWeight','Bold');
xlabel('Номер окна','FontSize',12); 

